content/asm_7.md

```diff
@@ -1,153 +1,268 @@
+This is straightforward C code which I think not needed to be explained in details. It checks that exactly one command-line argument was passed, then calls the `my_strlen` function to determine its length. The result is printed to standard output. As I said - pretty straightforward. The single part which interests us is the implementation of the `my_strlen` function. This function is written in assembly, but not in C.
+
+Let's take a look at the implementation of this function:
 
+```assembly
+;; Definition of the text section
+section .text
+        ;; Reference to the entry point of our program
+        global my_strlen
+
+;; Function that returns the length of the string passed via the first argument
+my_strlen:
+        ;; Reset the register value to zero. It will be returned from the
+        ;; function with the result.
+        xor rax, rax
+.loop:
+        ;; Compare the first element in the given string with the `NUL` terminator (end of the string).
+        cmp byte [rdi + rax], 0
+        ;; If we reached the `NUL` terminator exit from the function.
+        je .done
+        ;; Increase the counter which stores the length of the string.
+        inc rax
+        ;; Repeat the operations above while we didn't reach the end of the string.
+        jmp .loop
+.done:
+        ;; Exit from the function and return the result in the `rax` register.
+        ret
 ```
-    This helps GCC to distinguish between the operands and registers. operands have a single % as prefix
+
+The way function's parameters are passed to our assembly function follows the same calling convention which we have seen in the section above. The first argument, which our case is a pointer to the string, passed in the `rdi` register. At the start of the function we clear the value of the `rax` register, since this will serve as our counter. Each time we encounter a character in the string, the loop increments `rax` using it as an accumulator for the string’s length. The loop continues character by character until it reaches the `NUL` terminator, which signals the end of the string. At that point, the value stored in rax is returned to the caller as the final result.
+
+We can build our program using the following commands:
+
+```bash
+nasm -f elf64 -o casm.o casm.asm
+gcc casm.o casm.c -o casm
 ```
 
-We can build it with:
+And run it:
 
 ```
-build:
-	gcc casm.c -o casm
+./casm hello
+The argument length is - 5
 ```
 
-## Call C from assembly
+## Use inline assembly in C code
+
+In the section above, we already have mentioned that calling an assembly function from a C code is not a single way to use assembly code from within a C program. The second method is to use `inline assembly`. This method allows us to write an assembly code directly in our C code. This can be useful for performance-critical operations, or when you need to access processor instructions that the compiler does not expose through standard C.
```

Comment:
```suggestion
In the section above, we already mentioned that calling an assembly function from C code is not the only way to use assembly code from within a C program. The second method is to use inline assembly. This method allows us to write assembly code directly in our C code. This can be useful for performance-critical operations, or when you need to access processor instructions that the compiler does not expose through standard C.
```